File: js/obfuscator.py
import esprima
import escodegen
import base64
import random
import utils
from urllib.parse import quote

all_encode_string = ["ZGVjb2RlVVJJQ29tcG9uZW50KCI="]
inner_func_and_var = {}
all_identifier = {}
func_args = {}
instructions = []


def add_decode_function(ast):
    decode_unescape_function = esprima.parseScript(
        'function decodeUnescape(value){return eval(decodeBase64(0) + value + "\\")");}'
    )
    ast.body.insert(0, decode_unescape_function.body[0])

    decode_base64_function = esprima.parseScript(
        'function decodeBase64(index){return eval(' +
        '"d4e454c33o44d21e21U2R50I1C42o13m2p49o18n9e21n17t16(8e49s2c27a2p8e41(33w3i35n4d25o25w19.2a17t31o50b438(4\\""' +
        '.replace(/[0-9]/g,"") + all_string[index] + "\\")))");}'
    )
    ast.body.insert(0, decode_base64_function.body[0])

    decode_replace_encoding = esprima.parseScript(
        'function decodeReplace(value, r){return value.replaceAll(r,  "")}'
    )
    ast.body.insert(0, decode_replace_encoding.body[0])

    decode_reverse = esprima.parseScript(
        'function rev(value){ var a=value.split(""); var b=a.reverse(); var c=b.join("");return c;}'
    )
    ast.body.insert(0, decode_reverse.body[0])


def add_decode_array(ast):
    encode_string_array = []
    for s in all_encode_string:
        encode_string_array.append(esprima.nodes.Literal(s, f'"{s}"'))

    left = esprima.nodes.Identifier("all_string")
    right = esprima.nodes.ArrayExpression(encode_string_array)
    expression = esprima.nodes.AssignmentExpression("=", left, right)
    ast.body.insert(0, esprima.nodes.ExpressionStatement(expression))


def gen_expression(var_names):
    opt = random.randint(1 if len(var_names) > 0 else 2, 3)
    if opt == 1:
        return esprima.nodes.Identifier(random.choice(var_names))
    elif opt == 2:
        val = random.randint(1, 1000)
        return esprima.nodes.Literal(val, f'{val}')
    else:
        val = utils.generate_unique_sequence(random.randint(1, 10))
        return esprima.nodes.Literal(val, f'"{val}"')


def traverse(node, func_before=None, func_after=None):
    if func_before is not None:
        new_node = func_before(node)
        if new_node != node:
            return new_node

    node_items = node.items()
    for key, value in node_items:
        if key != 'type':
            if isinstance(value, list):
                for i, item in enumerate(value):
                    value[i] = traverse(item, func_before, func_after)
            elif isinstance(value, esprima.nodes.Node):
                setattr(node, key, traverse(value, func_before, func_after))

    if func_after is not None:
        new_node = func_before(node)
        if new_node != node:
            return new_node

    return node


def encode_string(node):
    def base64_encoding(value):
        encoded_value = base64.b64encode(value.encode('utf-8')).decode('utf-8')
        try:
            index = all_encode_string.index(encoded_value)
        except ValueError:
            all_encode_string.append(encoded_value)
            index = len(all_encode_string) - 1

        return esprima.nodes.CallExpression(
            esprima.nodes.Identifier("decodeBase64"),
            [esprima.nodes.Literal(index, f'"{len(all_encode_string) - 1}"')],
        )

    def unescape_encoding(value):
        # val = ''.join([f"%{ord(char):02x}" for char in value])
        val = quote(value, safe='')

        return esprima.nodes.CallExpression(
            esprima.nodes.Identifier("decodeUnescape"),
            [esprima.nodes.Literal(val, f'"{val}"')],
        )

    def replace_encoding(value):
        while True:
            random_sequence = utils.generate_unique_sequence(random.randint(3, 5))
            if random_sequence not in value:
                break

        parts = utils.get_parts(value)
        if len(parts) == 1:
            index_to_insert = int(len(value)/2)
            val = value[:index_to_insert] + random_sequence + value[index_to_insert:]
        else:
            val = random_sequence.join(parts)

        return esprima.nodes.CallExpression(
            esprima.nodes.Identifier("decodeReplace"),
            [esprima.nodes.Literal(val, f'"{val}"'), esprima.nodes.Literal(random_sequence, f'"{random_sequence}"')],
        )

    def reverse_encoding(value):
        val = value[::-1]

        return esprima.nodes.CallExpression(
            esprima.nodes.Identifier("rev"),
            [esprima.nodes.Literal(val, f'"{val}"')],
        )

    def encode_string_literal(node):
        if node.type == 'Literal' and isinstance(node.value, str):
            opt = random.randint(1, 4)
            if opt == 1:
                return base64_encoding(node.value)
            elif opt == 2:
                return unescape_encoding(node.value)
            elif opt == 3:
                return replace_encoding(node.value)
            elif opt == 4:
                return reverse_encoding(node.value)

        return node

    traverse(node, func_before=encode_string_literal)


def rename_identifier(node):
    def rename(node):
        if node.type == 'Identifier' and node.name in inner_func_and_var:
            if node.name not in all_identifier:
                all_identifier[node.name] = utils.gen_name(all_identifier)

            node.name = all_identifier[node.name]
        return node

    traverse(node, func_before=rename)


def collect_identifier(node):
    def collect(node):
        if node.type == 'FunctionDeclaration':
            inner_func_and_var[node.id.name] = {}
            for p in node.params:
                inner_func_and_var[p.name] = {}
        elif node.type == 'AssignmentExpression':
            inner_func_and_var[node.left.name] = {}
        elif node.type == 'VariableDeclarator':
            inner_func_and_var[node.id.name] = {}

        return node

    traverse(node, func_before=collect)


def collect_instruction(node):
    def collect(node):
        if node.type == 'ExpressionStatement':
            instructions.append(node)
        elif node.type == 'ForStatement':
            instructions.append(node)
        elif node.type == 'IfStatement':
            instructions.append(node)
        elif node.type == 'WhileStatement':
            instructions.append(node)

        return node

    traverse(node, func_before=collect)


def transform_constants(node):
    def get_literal(number):
        literal = esprima.nodes.Literal(abs(number), f'{abs(number)}')
        if number > 0:
            return literal
        return esprima.nodes.UnaryExpression('-', literal)

    def transform(node):
        if node.type == 'Literal' and isinstance(node.value, int):
            opt = random.randint(1, 5)
            if opt == 1:
                number1 = random.randint(1, node.value + 1)
                if number1 < node.value:
                    number2 = number1 - node.value

                    return esprima.nodes.BinaryExpression(
                        '-',
                        get_literal(number1),
                        get_literal(number2),
                    )
                else:
                    number2 = node.value - number1

                    return esprima.nodes.BinaryExpression(
                        '+',
                        get_literal(number2),
                        get_literal(number1),
                    )
            elif opt == 2:
                number1 = random.randint((node.value + 1), (node.value + 1) * 10)
                number2 = number1 * random.randint(1, 100) + node.value

                return esprima.nodes.BinaryExpression(
                    '%',
                    get_literal(number2),
                    get_literal(number1),
                )
            elif opt == 3:
                number1 = random.randint(1, (node.value + 1) * 10)
                number2 = node.value ^ number1

                return esprima.nodes.BinaryExpression(
                    '^',
                    get_literal(number2),
                    get_literal(number1),
                )

        return node

    for i in range(random.randint(1, 3)):
        node = traverse(node, func_before=transform)
    return node


def split_string(node):
    def split(node):
        if node.type == 'Literal' and isinstance(node.value, str):
            parts = utils.get_parts(node.value)
            if len(parts) == 1:
                return node

            left = esprima.nodes.Literal(parts[0], f'"{parts[0]}"')
            for part in parts[1:]:
                left = esprima.nodes.BinaryExpression('+', left, esprima.nodes.Literal(part, f'"{part}"'))
            return left

        return node

    traverse(node, func_before=split)


def add_args(node):
    def add_to_declaration(node):
        if node.type == 'FunctionDeclaration':
            new_count_arg = random.randint(len(node.params), len(node.params)*2 + 3)
            idx = utils.generate_unique_random_numbers(len(node.params), new_count_arg)

            func_args[node.id.name] = {'idx': idx, 'count': new_count_arg}

            params = []
            for i in range(new_count_arg):
                params.append(esprima.nodes.Identifier(utils.generate_unique_sequence(3)))

            for index, value in enumerate(node.params):
                params[idx[index]] = value

            node.params = params
        return node

    def add_to_call(node):
        if node.type == 'CallExpression':
            if node.callee.name not in func_args:
                return node

            args_name = []
            for a in node.arguments:
                if a.type == 'Identifier':
                    args_name.append(a.name)

            new_args = func_args[node.callee.name]
            if len(node.arguments) == new_args['count']:
                return node

            args = []
            for i in range(new_args['count']):
                val = gen_expression(args_name)
                args.append(val)

            for index, value in enumerate(node.arguments):
                args[new_args['idx'][index]] = value

            node.arguments = args

        return node

    traverse(node, func_before=add_to_declaration)
    traverse(node, func_before=add_to_call)


def add_fake_function(ast):
    def gen_body():
        body = []
        for _ in range(random.randint(1, min(len(instructions), 15))):
            body.append(random.choice(instructions))
        return esprima.nodes.BlockStatement(body)

    def gen_args():
        return []

    def gen_func():
        func_name = utils.gen_name(all_identifier)
        body = gen_body()
        args = gen_args()

        return esprima.nodes.FunctionDeclaration(
            esprima.nodes.Identifier(func_name),
            args,
            body,
            False
        )

    collect_instruction(ast)

    for _ in range(random.randint(1, min(len(ast.body), 5))):
        ast.body.insert(random.randint(0, len(ast.body) - 1), gen_func())


def obfuscate_code(code):
    ast = esprima.parseScript(code)
    print(ast)

    split_string(ast)
    encode_string(ast)
    add_decode_array(ast)
    add_decode_function(ast)
    add_fake_function(ast)
    add_args(ast)
    collect_identifier(ast)
    rename_identifier(ast)
    transform_constants(ast)
    #print(ast)

    return escodegen.generate(ast, options={
        'format': {'indent': {'style': ''}, 'newline': '', 'space': '', 'hexadecimal': True}})


if __name__ == '__main__':
    file_path = ""
    file_path = "test.js"

    input_code = """
        function add(a, b) {
            return a + b;
        }
        add("a", "b")
    """

    if file_path != "":
        with open(file_path, 'r', encoding='utf-8') as file:
            input_code = file.read()

    obfuscated_code = obfuscate_code(input_code)

    with open('res.js', 'w') as file:
        file.write(obfuscated_code)

    print("\n\n\n")
    #print(obfuscated_code)

File: ruby/obfuscator.rb
require 'parser/current'
require 'unparser'
require 'base64'

$inner_func_and_var = {}
$all_identifier = {}

def extract_prefix(str)
  match = str.match(/^(@@|@|\$)/)
  match.nil? ? "" : match[0]
end

def gen_name
  name = "_0ib#{rand(10..999)}k#{rand(1000..9999)}s"

  while $all_identifier.key?(name)
    name = "_0ib#{rand(10..999)}k#{rand(1000..9999)}s"
  end

  name
end

def get_expr(operation, left, right)
  Parser::AST::Node.new(
    :begin,
    [
      Parser::AST::Node.new(
        :send,
        [
          Parser::AST::Node.new(:int, [left]),
          operation.to_sym,
          Parser::AST::Node.new(:int, [right])]
        )]
    )
end

def get_parts(input_string)
  return [input_string] if input_string.length == 1

  num_parts = rand(1..(input_string.length / 2) + 1)
  return [input_string] if num_parts == 1

  split_indices = (1..input_string.length - 1).to_a.sample(num_parts - 1).sort

  parts = split_indices.each_with_index.map do |j, i|
    prev_index = i.zero? ? 0 : split_indices[i - 1] + 1
    input_string[prev_index..j]
  end

  parts << input_string[split_indices.last + 1..-1] if split_indices.last
  parts
end

def get_base64_encoding(string)
  args = [nil, :"decode_base64", Parser::AST::Node.new(:str, [base64(string).to_s])]
  Parser::AST::Node.new(:send, args)
end

def base64(input_string)
  encoded_string = Base64.encode64(input_string)
  encoded_string.chomp
end


def traverse(node, skip_new, &func_before)
  if func_before.is_a?(Proc)
    new_node = func_before.call(node)
    if new_node != node && skip_new
      return new_node
    end
  end

  new_child = new_node.children.dup
  new_child.each_with_index do |value, index|
    if value.is_a?(Parser::AST::Node)
      new_subnode = traverse(value, skip_new, &func_before)
      if new_subnode != new_node
        new_child[index] = new_subnode
      end
    end
  end
  new_node = new_node.updated(new_node.type, new_child)

  new_node
end

def collect_identifier(node)
  collect = lambda do |coll_node|
    types0 = [:lvasgn, :gvasgn, :cvasgn, :ivasgn, :arg, :def]
    types1 = [:const]

    if types0.include?(coll_node.type)
      $inner_func_and_var[coll_node.children[0]] = {}
    elsif types1.include?(coll_node.type) && coll_node.children[1] != :Base64
      $inner_func_and_var[coll_node.children[1]] = {}
    end

    return coll_node
  end

  traverse(node, false, &collect)
end


def rename_identifier(node)
  rename = lambda do |rename_node|
    types0 = [:lvasgn, :gvasgn, :cvasgn, :ivasgn, :lvar, :arg, :ivar, :cvar, :gvar, :def]
    types1 = [:send, :const]

    if types0.include?(rename_node.type) && $inner_func_and_var.key?(rename_node.children[0])
      unless $all_identifier.key?(rename_node.children[0])
        prefix = extract_prefix(rename_node.children[0].name)
        $all_identifier[rename_node.children[0]] = prefix + gen_name
      end


      new_child = rename_node.children.dup
      new_child[0] = $all_identifier[rename_node.children[0]].to_sym
      rename_node = rename_node.updated(rename_node.type, new_child)
    elsif types1.include?(rename_node.type) && $inner_func_and_var.key?(rename_node.children[1])
      unless $all_identifier.key?(rename_node.children[1])
        prefix = extract_prefix(rename_node.children[1].name)
        if rename_node.type == :const
          prefix = 'C' + prefix
        end
        $all_identifier[rename_node.children[1]] = prefix + gen_name
      end


      new_child = rename_node.children.dup
      new_child[1] = $all_identifier[rename_node.children[1]].to_sym
      rename_node = rename_node.updated(rename_node.type, new_child)
    end

    return rename_node
  end

  traverse(node, false, &rename)
end


def transform_constants(node)
  transform = lambda do |transform_node|
    if transform_node.type == :int && transform_node.children.length == 1
      opt = rand(1..4)

      value = transform_node.children[0].to_s.to_i
      case opt
      when 1
        number1 = rand(1..(value + 1)*2)
        if value < number1
          number2 = number1 - value
          return get_expr("-", number1, number2)
        else
          number2 = value - number1
          return get_expr("+", number1, number2)
        end
      when 2
        number1 = rand((value + 1)..((value + 1) * 10))
        number2 = number1 * rand(1..100) + value
        return get_expr("%", number2, number1)
      when 3
        number1 = rand(1..((value + 1) * 10))
        number2 = value ^ number1
        return get_expr("^", number2, number1)
      end
    end

    return transform_node
  end


  for i in 1..4
    node = traverse(node,true, &transform)
  end

  node
end

def split_string(node)
  split = lambda do |split_node|
    if split_node.type == :str
      parts = get_parts(split_node.children[0].to_s)
      if parts.length == 1
        return split_node
      end

      left = Parser::AST::Node.new(:str, [parts[0].to_s])
      parts[1..-1].each do |part|
        left = Parser::AST::Node.new(:send, [left, :+, Parser::AST::Node.new(:str, [part])])
      end
      return left
    end

    split_node
  end

  traverse(node, true, &split)
end

def encoding_string(node)
  encoding = lambda do |encoding_node|
    if encoding_node.type == :str
      opt = rand(1..2)
      if opt == 1
        value = encoding_node.children[0].to_s
        return get_base64_encoding(value)
      end

      return encoding_node
    end

    encoding_node
  end

  traverse(node, true, &encoding)
end

def add_decode_function(node)
  f = Unparser.parse("def decode_base64(encoded_string)\nBase64.decode64(encoded_string)\nend")

  new_child = node.children.dup
  new_child.insert(0, f)

  node.updated(node.type, new_child)
end

def add_require(node)
  req = Unparser.parse("require 'base64'")
  new_child = node.children.dup
  new_child.insert(0, req)

  node.updated(node.type, new_child)
end

input_filename = ARGV[0]
output_filename = ARGV[1]

input_filename = "test.rb"
output_filename = "res.rb"

if input_filename.nil? || output_filename.nil?
  puts "Usage: ruby read_file.rb <input_filename> <output_filename>"
  exit
end

begin
  file_content = File.read(input_filename)
rescue Errno::ENOENT
  abort("Error: File '#{filename}' not found.")
rescue Errno::EACCES
  abort("Error: Permission denied for file '#{filename}'.")
rescue => e
  abort("An unexpected error occurred: #{e.message}")
end

ast = Unparser.parse(file_content)
ast = ast.dup
puts ast

ast = add_decode_function(ast)
ast = split_string(ast)
ast = encoding_string(ast)
ast = collect_identifier(ast)
ast = rename_identifier(ast)
ast = transform_constants(ast)
ast = add_require(ast)



#puts $inner_func_and_var
result_content = Unparser.unparse(ast)
puts result_content
File.open(output_filename, 'w') { |file| file.write(result_content) }
